<a href="https://colab.research.google.com/github/atoothman/DATA-70500/blob/main/Final_Lab_4__LinearModels_Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Linear Models, part II: Logistic Regression

We'll use the global social indicators data to develop a logistic regression model and pratice interpreting the results.

First, we'll import the libraries we'll need for this model.


In [62]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import seaborn as sb
import math


Here's some information about where these social indicators were created: https://hdr.undp.org/data-center/composite-indices


Next, we'll read in the data sources and create the DataFrame with all of our variables.

In [ ]:
GlobalIndicators1 = pd.read_excel('http://data.shortell.nyc/files/HumanDevelopment.xlsx', index_col='Country', na_values=[np.nan])
GlobalIndicators1.head()

HDI Rank  Human Development Index (HDI)  \
Country                                                
Norway              1                          0.944   
Australia           2                          0.935   
Switzerland         3                          0.930   
Denmark             4                          0.923   
Netherlands         5                          0.922   

             Life Expectancy at Birth  Expected Years of Education  \
Country                                                              
Norway                           81.6                         17.5   
Australia                        82.4                         20.2   
Switzerland                      83.0                         15.8   
Denmark                          80.2                         18.7   
Netherlands                      81.6                         17.9   

             Mean Years of Education  Gross National Income (GNI) per Capita  \
Country                                                                        
Norway                          12.6                                   64992   
Australia                       13.0                                   42261   
Switzerland                     12.8                                   56431   
Denmark                         12.7                                   44025   
Netherlands                     11.9                                   45435   

             GNI per Capita Rank Minus HDI Rank  
Country                                          
Norway                                        5  
Australia                                    17  
Switzerland                                   6  
Denmark                                      11  
Netherlands                                   9

In [ ]:
GlobalIndicators2 = pd.read_excel('http://data.shortell.nyc/files/GenderDevelopment.xlsx', index_col='Country', na_values=[np.nan])
GlobalIndicators2.head()

GDI Rank  Gender Development Index (GDI)  \
Country                                                 
Norway              1                           0.996   
Australia           2                           0.976   
Switzerland         3                           0.950   
Denmark             4                           0.977   
Netherlands         5                           0.947   

             Human Development Index (Female)  Human Development Index (Male)  \
Country                                                                         
Norway                                  0.940                           0.944   
Australia                               0.922                           0.945   
Switzerland                             0.898                           0.945   
Denmark                                 0.912                           0.934   
Netherlands                             0.893                           0.943   

             Life Expectancy at Birth (Female)  \
Country                                          
Norway                                    83.6   
Australia                                 84.5   
Switzerland                               85.0   
Denmark                                   82.2   
Netherlands                               83.3   

             Life Expectancy at Birth (Male)  \
Country                                        
Norway                                  79.5   
Australia                               80.3   
Switzerland                             80.8   
Denmark                                 78.3   
Netherlands                             79.7   

             Expected Years of Education (Female)  \
Country                                             
Norway                                       18.2   
Australia                                    20.7   
Switzerland                                  15.7   
Denmark                                      19.3   
Netherlands                                  18.0   

             Expected Years of Education (Male)  \
Country                                           
Norway                                     16.8   
Australia                                  19.7   
Switzerland                                15.9   
Denmark                                    18.1   
Netherlands                                17.9   

             Mean Years of Education (Female)  Mean Years of Education (Male)  \
Country                                                                         
Norway                                   12.7                            12.5   
Australia                                13.1                            12.9   
Switzerland                              11.5                            13.1   
Denmark                                  12.8                            12.7   
Netherlands                              11.6                            12.2   

             Estimated Gross National Income per Capita (Female)  \
Country                                                            
Norway                                                 57140.0     
Australia                                              33688.0     
Switzerland                                            44132.0     
Denmark                                                36439.0     
Netherlands                                            29500.0     

             Estimated Gross National Income per Capita (Male)  
Country                                                         
Norway                                                 72825.0  
Australia                                              50914.0  
Switzerland                                            69077.0  
Denmark                                                51727.0  
Netherlands                                            61641.0

In [ ]:
GlobalIndicators3 = pd.read_excel('http://data.shortell.nyc/files/GenderInequality.xlsx', index_col='Country', na_values=[np.nan])
GlobalIndicators3.head()

GII Rank  Gender Inequality Index (GII)  \
Country                                                
Norway              1                          0.067   
Australia           2                          0.110   
Switzerland         3                          0.028   
Denmark             4                          0.048   
Netherlands         5                          0.062   

             Maternal Mortality Ratio  Adolescent Birth Rate  \
Country                                                        
Norway                            4.0                    7.8   
Australia                         6.0                   12.1   
Switzerland                       6.0                    1.9   
Denmark                           5.0                    5.1   
Netherlands                       6.0                    6.2   

             Percent Representation in Parliament  \
Country                                             
Norway                                       39.6   
Australia                                    30.5   
Switzerland                                  28.5   
Denmark                                      38.0   
Netherlands                                  36.9   

             Population with Secondary Education (Female)  \
Country                                                     
Norway                                               97.4   
Australia                                            94.3   
Switzerland                                          95.0   
Denmark                                              95.5   
Netherlands                                          87.7   

             Population with Secondary Education (Male)  \
Country                                                   
Norway                                             96.7   
Australia                                          94.6   
Switzerland                                        96.6   
Denmark                                            96.6   
Netherlands                                        90.5   

             Labour Force Participation Rate (Female)  \
Country                                                 
Norway                                           61.2   
Australia                                        58.8   
Switzerland                                      61.8   
Denmark                                          58.7   
Netherlands                                      58.5   

             Labour Force Participation Rate (Male)  
Country                                              
Norway                                         68.7  
Australia                                      71.8  
Switzerland                                    74.9  
Denmark                                        66.4  
Netherlands                                    70.6

In [ ]:
GlobalIndicatorsTotal = pd.concat([GlobalIndicators1, GlobalIndicators2, GlobalIndicators3], axis=1)
GlobalIndicatorsTotal.info('verbose')

<class 'pandas.core.frame.DataFrame'>
Index: 188 entries, Norway to Niger
Data columns (total 28 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   HDI Rank                                             188 non-null    int64  
 1   Human Development Index (HDI)                        188 non-null    float64
 2   Life Expectancy at Birth                             188 non-null    float64
 3   Expected Years of Education                          188 non-null    float64
 4   Mean Years of Education                              188 non-null    float64
 5   Gross National Income (GNI) per Capita               188 non-null    int64  
 6   GNI per Capita Rank Minus HDI Rank                   188 non-null    int64  
 7   GDI Rank                                             188 non-null    int64  
 8   Gender Development Index (GDI)                       161 non-null   

Now, we'll compute a binary variable that will be our dependent variable, Y. Then, we'll identify the relevant independent variables and put them in a new DataFrame, X. At that point, we can compute the model.

In [ ]:
GlobalIndicatorsTotal['Gender Inequality Index (GII)'].describe()

count    155.000000
mean       0.365884
std        0.191457
min        0.016000
25%        0.184000
50%        0.385000
75%        0.524500
max        0.744000
Name: Gender Inequality Index (GII), dtype: float64

In [ ]:
GlobalIndicatorsTotal['GII Binary'] = 0
GlobalIndicatorsTotal.loc[GlobalIndicatorsTotal['Gender Inequality Index (GII)'] < 0.19, ['GII Binary']] = 1 #These are nations with a low gender inequality score--that is, the highest gender equality
GlobalIndicatorsTotal['GII Binary'].describe()

count    188.000000
mean       0.207447
std        0.406561
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        1.000000
Name: GII Binary, dtype: float64

In [ ]:
GlobalIndicatorsTotal['GII Binary']

Country
Norway                      1
Australia                   1
Switzerland                 1
Denmark                     1
Netherlands                 1
                           ..
Burundi                     0
Chad                        0
Eritrea                     0
Central African Republic    0
Niger                       0
Name: GII Binary, Length: 188, dtype: int64

In [ ]:
Y = GlobalIndicatorsTotal['GII Binary']
X = GlobalIndicatorsTotal[['Percent Representation in Parliament', 'Population with Secondary Education (Female)', 'Labour Force Participation Rate (Female)', 'Life Expectancy at Birth', 'Gross National Income (GNI) per Capita']]
model0 = sm.Logit(Y, X, missing='drop').fit()
print(model0.summary())

Optimization terminated successfully.
         Current function value: 0.321133
         Iterations 8
                           Logit Regression Results                           
Dep. Variable:             GII Binary   No. Observations:                  156
Model:                          Logit   Df Residuals:                      151
Method:                           MLE   Df Model:                            4
Date:                Fri, 11 Oct 2024   Pseudo R-squ.:                  0.4289
Time:                        16:37:06   Log-Likelihood:                -50.097
converged:                       True   LL-Null:                       -87.724
Covariance Type:            nonrobust   LLR p-value:                 1.760e-15
                                                   coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------
Percent Representation in Parliament    

In [ ]:
print(math.exp(model0.params[0]), math.exp(model0.params[1]), 1/math.exp(model0.params[2]), 1/math.exp(model0.params[3]), math.exp(1000*model0.params[4]))
# We need to exponentiate (or the take anti-logs of) the coefficients in order to interpret them as odds. For the negative coefficients, it is useful to take the inverse of the result
# and interpret it in the opposite direction (that is, the odds of not being in the high gender equality group). You can also change the increment of change in X
# as is the case here with the parameter for GNI per capita; I changed the increment to $1000 instead of $1.

1.113047941683358 1.0677546839634509 1.0600888084476103 1.086297381643493 1.0540509355600345


<ipython-input-10-1b1cc13454ea>:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(math.exp(model0.params[0]), math.exp(model0.params[1]), 1/math.exp(model0.params[2]), 1/math.exp(model0.params[3]), math.exp(1000*model0.params[4]))


In [ ]:
print(np.exp(model0.params)) #These are expressed as odds ratios

Percent Representation in Parliament            1.113048
Population with Secondary Education (Female)    1.067755
Labour Force Participation Rate (Female)        0.943317
Life Expectancy at Birth                        0.920558
Gross National Income (GNI) per Capita          1.000053
dtype: float64


In [ ]:
model0_marginals = model0.get_margeff() #These are the average effects
print(model0_marginals.summary())

        Logit Marginal Effects       
Dep. Variable:             GII Binary
Method:                          dydx
At:                           overall
                                                  dy/dx    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------
Percent Representation in Parliament             0.0109      0.002      4.922      0.000       0.007       0.015
Population with Secondary Education (Female)     0.0066      0.001      5.889      0.000       0.004       0.009
Labour Force Participation Rate (Female)        -0.0059      0.002     -2.653      0.008      -0.010      -0.002
Life Expectancy at Birth                        -0.0084      0.001     -6.576      0.000      -0.011      -0.006
Gross National Income (GNI) per Capita        5.335e-06   1.13e-06      4.728      0.000    3.12e-06    7.55e-06


In [ ]:
model0_marginals = model0.get_margeff(at='median') #It is often more useful to get estimages of the effect sizes at particular values for the factors
print(model0_marginals.summary())

        Logit Marginal Effects       
Dep. Variable:             GII Binary
Method:                          dydx
At:                            median
                                                  dy/dx    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------
Percent Representation in Parliament             0.0055      0.002      2.711      0.007       0.002       0.009
Population with Secondary Education (Female)     0.0034      0.001      3.074      0.002       0.001       0.006
Labour Force Participation Rate (Female)        -0.0030      0.001     -2.608      0.009      -0.005      -0.001
Life Expectancy at Birth                        -0.0043      0.002     -2.647      0.008      -0.007      -0.001
Gross National Income (GNI) per Capita        2.708e-06    1.1e-06      2.458      0.014    5.48e-07    4.87e-06


In [ ]:
mdn_rep = np.exp(0.0055*10)
print("The effect of median representation in parliament is", f"{mdn_rep:.3f} times more likely to be a high quality nation for an increase of ten percent in representation.")

The effect of median representation in parliament is 1.057 times more likely to be a high quality nation for an increase of ten percent in representation.


In [ ]:
model0_pred = model0.pred_table()
print(model0_pred) # Correct predictions are on the diagonal of the 2d array.

[[109.   8.]
 [ 12.  27.]]


In [ ]:
correct_i = 109 / (109 + 8) # The proportion of correct predictions of 0.
correct_j = 27 / (27 + 12) # The proportion of correct predictions of 1.
print(correct_i, correct_j)

0.9316239316239316 0.6923076923076923


## Activity

1. Find and read into a DataFrame a suitable dataset. You may use the global social indictors data from the example here. You may need to combine files, as shown here.

2. Identify a dependent variable to explain. Create a binary variable of this measure, if needed. Explain why you chose this variable or recoded in the way you did.

3. Build a model to explain the DV. You can use the odds (anti-log of the coefficients) or the marginal effects to test for the unique effects of each predictor.

4. Explain the results.

In [63]:
#1 Find and read into a DataFrame a suitable dataset. You may use the global social indictors data from the example here. You may need to combine files, as shown here.

import pandas as pd

# Saved url on github of pima indians bmi
url = 'https://raw.githubusercontent.com/atoothman/DATA-70500/main/Predict_BMI.csv'

# Load the CSV file into a dataframe
df = pd.read_csv(url)

# Display the dataframe
df.head()

Height M  Weight kg        BMI  Percent Fat
0   1.60020  49.441572  19.308287         23.9
1   1.65100  62.595751  22.964168         28.8
2   1.65100  75.749931  27.789971         32.4
3   1.53035  48.987979  20.917414         25.8
4   1.45415  43.091278  20.378441         22.5

In [64]:
#2 Identify a dependent variable to explain. Create a binary variable of this measure, if needed. Explain why you chose this variable or recoded in the way you did.

# Will select BMI as the dependant variable to explain. BMi is calcualted based off height and weight so it is dependant on these factors. It will need to be converted into a binary variable 0 = not obese and 1 obese.
# It needs to be convereted tto a binary variable becuase it is current a continue variable which will not work in the model

print(df['BMI'].head())

0    19.308287
1    22.964168
2    27.789971
3    20.917414
4    20.378441
Name: BMI, dtype: float64


In [65]:
# Convert 'BMI' column to binary: 1 for BMI >= 30, 0 for BMI < 30
df['BMI'] = df['BMI'].apply(lambda x: 1 if x >= 25 else 0)

# Check the result
print(df[['BMI']].head())

   BMI
0    0
1    0
2    1
3    0
4    0


In [66]:
# Count how many 'BMI' values are 1 (overweight)
count_overweight = df['BMI'].sum()

print('Total number of BMI = 1 (Overweight):', count_overweight)

Total number of BMI = 1 (Overweight): 13


In [67]:
#3. Build a model to explain the DV. You can use the odds (anti-log of the coefficients) or the marginal effects to test for the unique effects of each predictor.

# Import stats model for log regression model and for adding constant below
import statsmodels.api as sm

# Add a constant (intercept) to the independent variables. This was not working in subsequent steps without adding it.
X = sm.add_constant(df[['Height M', 'Percent Fat']])

# Fit the logistic regression model
modelBMI = sm.Logit(df['BMI'], X, missing='drop').fit()

# Print the model summary
print(modelBMI.summary())


Optimization terminated successfully.
         Current function value: 0.127754
         Iterations 10
                           Logit Regression Results                           
Dep. Variable:                    BMI   No. Observations:                   92
Model:                          Logit   Df Residuals:                       89
Method:                           MLE   Df Model:                            2
Date:                Fri, 11 Oct 2024   Pseudo R-squ.:                  0.6864
Time:                        19:41:17   Log-Likelihood:                -11.753
converged:                       True   LL-Null:                       -37.474
Covariance Type:            nonrobust   LLR p-value:                 6.757e-12
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
const         -43.5113     16.584     -2.624      0.009     -76.016     -11.007
Height M       15.1650   

In [68]:
# Need to take the logs of the coef Height and Weight to interpret
import math

# print(math.exp(modelBMI.params[0]), math.exp(modelBMI.params[1]), 1/math.exp(modelBMI.params[2]), math.exp(1000*modelBMI.params[2]))
# We need to exponentiate (or the take anti-logs of) the coefficients in order to interpret them as odds. For the negative coefficients, it is useful to take the inverse of the result
# and interpret it in the opposite direction (that is, the odds of not being in the high gender equality group). You can also change the increment of change in X
# as is the case here with the parameter for GNI per capita; I changed the increment to $1000 instead of $1.


# Calculating odd ratios
print(
    math.exp(modelBMI.params[0]),  # Exponentiated intercept
    math.exp(modelBMI.params[1]),  # Exponentiated coefficient for Height
    math.exp(modelBMI.params[2])   # Exponentiated coefficient for Weight
)

1.2684790166028542e-19 3855291.957009628 1.7405482694074603


<ipython-input-68-b82976275886>:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  math.exp(modelBMI.params[0]),  # Exponentiated intercept
<ipython-input-68-b82976275886>:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  math.exp(modelBMI.params[1]),  # Exponentiated coefficient for Height
<ipython-input-68-b82976275886>:12: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  math.exp(modelBMI.params[2])   # Exponentiated coefficient for 

In [69]:
print(np.exp(modelBMI.params)) #These are expressed as odds ratios

const          1.268479e-19
Height M       3.855292e+06
Percent Fat    1.740548e+00
dtype: float64


4. Explain the results.

This logistic regression model describes the probability of a person being overweight or not overweight. The BMI is the dependent variable and was converted into a binary variable for this analysis, with overweight coded as 1 and not overweight as 0. The independent variables are Height and Fat Percentage.

When examining the coefficient of the model, Height is 15.2, indicating that as height increases, a person is more likely to be obese. When the coefficient is exponentiated, it returns 3.855292e+06, suggesting that the taller the individual, the more likely they are to be overweight. This number is so large that it is suspicious, and multicollinearity should be considered.

However, the coefficient for Percent Fat is 0.55, and when it is exponentiated, it is 1.740548e+00. When the percentage odds are calculated, we can see that there is a 74% increase in odds. This indicates that for each additional point of body fat, the individual's likelihood of being overweight increases by 74%.